src/api/script-cache.ts

```diff
@@ -0,0 +1,106 @@
+import { Link } from "expression/link";
+import { Datastore } from "index/datastore";
+import { Result } from "./result";
+import { MarkdownCodeblock, MarkdownSection } from "index/types/markdown";
+import { DatacoreJSRenderer, ScriptLanguage, asyncEvalInContext } from "ui/javascript";
+import { DatacoreLocalApi } from "./local-api";
+import { Fragment, h } from "preact";
+export interface DatacoreScript {
+    language: ScriptLanguage;
+    code: string | null;
+    id: string;
+    state: LoadingState;
+    source: string;
+}
+export const enum LoadingState {
+    UNDEFINED = -1,
+    LOADING,
+    LOADED,
+}
+/** A simple caching script loader that can load any DAG of script dependencies. */
+export class ScriptCache {
+    /** A static placeholder placed into the `scripts` section to catch recursive loading loops. */
+    private static readonly LOADING_SENTINEL: unique symbol = Symbol("CURRENTLY_LOADING");
+    private static SCRIPT_TAGS: string[] = ["javascript", "typescript", "ts", "js", "tsx", "jsx"];
+
+    public constructor(private store: Datastore) {}
+    /**
+     * Load the given script at the given path, recursively loading any subscripts by correctly injecting `import` into the child contexts.
+     */
+    public async load(
+        path: string | Link,
+        parentContext: DatacoreLocalApi,
+        sourcePath?: string
+    ): Promise<Result<any, string>> {
+        const source = await this.resolveSource(path, sourcePath);
+        if (!source.successful) return source;
+        source.value.state = LoadingState.LOADING;
+        let element = parentContext.loadedScripts.find(
+            (x) => x.id === source.value.id && x.state === LoadingState.LOADING
+        );
+        if (!element) {
+            parentContext.loadedScripts.push(source.value);
+            let dc = new DatacoreLocalApi(parentContext.api, source.value.source, parentContext.loadedScripts);
+            const res = await asyncEvalInContext(source.value.code!, {

```

Comment:
Why does `code` need to be nullable? You only need it locally once when you load the contents so you can execute them.

Reply:
I see you are doing this presumably to hold onto `code` longer than needed. I would do this by separating the type for "resolving the source" and "the resolved script" so you don't hold onto the script metadata longer than needed.